Mathlib/CategoryTheory/WithTerminal/Cone.lean

```diff
@@ -0,0 +1,127 @@
+# Relations between `Cone`, `WithTerminal` and `Over`
+
+Given categories `C` and `J`, an object `X : C` and a functor `K : J ⥤ Over X`,
+it has an obvious lift `liftFromOver K : WithTerminal J ⥤ C`, namely, send the terminal
+object to `X`. These two functors have equivalent categories of cones (`coneEquiv`).
+As a corollary, the limit of `K` is the limit of `liftFromOver K`, and viceversa.
+-/
+
+open CategoryTheory CategoryTheory.Limits CategoryTheory.WithTerminal
+
+universe w w' v₁ v₂ u₁ u₂
+variable {C : Type u₁} [Category.{v₁} C]
+variable {D : Type u₂} [Category.{v₂} D]
+variable {J : Type w} [Category.{w'} J]
+
+namespace CategoryTheory.Limits.WithTerminal
+
+/-- The category of functors `J ⥤ Over X` can be seen as part of a comma category,
+namely the comma category constructed from the identity of the category of functors
+`J ⥤ C` and the functor that maps `X : C` to the constant functor `J ⥤ C`.
+
+Given a functor `K : J ⥤ Over X`, it is mapped to a natural transformation from the
+obvious functor `J ⥤ C` to the constant functor `X`. -/
+@[simps]
+def commaFromFunctorToOver {X : C} : (J ⥤ Over X) ⥤ Comma (𝟭 (J ⥤ C)) (Functor.const J) where
+  obj K := {
+    left := K ⋙ Over.forget X
+    right := X
+    hom.app a := (K.obj a).hom
+  }
+  map f := {
+    left := whiskerRight f (Over.forget X)
+    right := 𝟙 X
+  }
+
+/-- For any functor `K : J ⥤ Over X`, there is a canonical extension
+`WithTerminal J ⥤ C`, that sends `star` to `X`. -/
+@[simps!]
+def liftFromOverToWithTerminal {X : C} : (J ⥤ Over X) ⥤ WithTerminal J ⥤ C :=
+  commaFromFunctorToOver ⋙ equivComma.inverse
+
+/-- The extension of a functor to over categories behaves well with compositions. -/
+@[simps]
+def extendCompose {X : C} (K : J ⥤ Over X) (F : C ⥤ D) :
+    liftFromOverToWithTerminal.obj K ⋙ F ≅
+    liftFromOverToWithTerminal.obj (K ⋙ Over.post F) where
+  hom.app
+  | star => 𝟙 _
+  | of a => 𝟙 _
+  inv.app
+  | star => 𝟙 _
+  | of a => 𝟙 _
+
+/-- A cone of a functor `K : J ⥤ Over X` consists of an object of `Over X`, together
+with morphisms. This same object is a cone of the extended functor
+`liftFromOver.obj K : WithTerminal J ⥤ C`. -/
+@[simps]
+private def coneLift {X : C} {K : J ⥤ Over X} :
+    Cone K ⥤ Cone (liftFromOverToWithTerminal.obj K) where
+  obj t := {
+    pt := t.pt.left
+    π.app
+    | of a => (t.π.app a).left
+    | star => t.pt.hom
+    π.naturality
+    | star, star, _
+    | of a, star, _ => by aesop
+    | star, of _, _ => by contradiction
+    | of a, of b, f => by simp [← Comma.comp_left]
+  }
+  map {t₁ t₂} f := {
+    hom := f.hom.left
+    w
+    | star => by aesop_cat
+    | of a => by simp [← Comma.comp_left]
+  }
+
+/-- This is the inverse of the previous construction: a cone of an extended functor
+`liftFromOver.obj K : WithTerminal J ⥤ C` consists of an object of `C`, together
+with morphisms. This same object is a cone of the original functor `K : J ⥤ Over X`. -/
+@[simps]
+private def coneBack {X : C} {K : J ⥤ Over X} :
```

Comment:
This should not be private if it appears in the simp lemma `coneEquiv_inverse`. Same for `coneLift`

Reply:
Hmmm the `!` didn't show up last time when I looked at the file but it is there now :magic:.
But I'm not sure if `coneEquiv_functor_obj_π_app` is a desired simp lemma.

Reply:
I don't know too much about that, but I guess it doesn't hurt? What simp lemmas do we even want here?

Reply:
I'm not sure `match` should be on the RHS of a simp lemma. Maybe add `coneEquiv_functor_obj_π_app_star` and `coneEquiv_functor_obj_π_app_of` instead?

Reply:
I guess you mean that we should write them explicitly. I don't think `simps` can be made to generate them, right?